Bounce them all!

Inky-Inky opened the pull request:
Any type of (player or monster) projectile can now bounce on a trampoline

Inky-Inky committed:
Update rm_mechanics.qc

Any type of (player or monster) projectile can now bounce on a trampoline

Inky-Inky committed:
Cushion only

Cushion only mode for trampolines

Inky-Inky committed:
Bounces count

Counting the number of bounces for each bouncing entity in case someone cares + fixing a bug with the bounceable check on top of trampoline_touch

Inky-Inky commented:
This PR was initially meant to be focused but eventually embarks 3 small new features:
- The trampoline is now able to be used simply as a cushion against fall damage (without bouncing effect) by means of a spawnflag.
- All kind of projectiles are now able to bounce: player grenades as before, but now also nails & rockets, and monsters projectiles as well
- Each bouncee keeps track of how many times they bounced thanks to its .maxbounce field (now able to be checked by trigger_filter)

EmeraldTiger reviewed (commented):
Cushion only spawnflag could be generalized to a "minbounce" field (the opposite of the existing maxbounce field). The value input into this would be the minimum speed the player must be falling to be bounced by the trampoline. 

By default the minspeed is 200 ups (abs value of what's defined in the code). If set to -1 this would have the same behavior as the existing cushion only spawnflag, while also allowing players to set their own minspeed as needed.

Inky-Inky committed:
No bouncing spikes

Prevent spikes from bouncing on trampolines (too odd)

Inky-Inky committed:
Cushion only version 2

Cushion-only behavior not triggered by spawnflag anymore but by setting speed to -1.
Also, minimum speed required to bounce now configurable by the mapper.

Inky-Inky committed:
Oooops (typo in the fgd file)

EmeraldTiger committed:
Revert to old bounce filters, keep cushions/maxbounce

With the current setup, only gravity-affected entities are eligible to be affected by trampolines, which is intuitive behavior, whereas rockets/nails, though fun for sure, are more questionable (why does the rocket bounce instead of just mangling the trampoline/destroying it?). The exclusion of nails only adds to the complexity of what is an eligible bouncee.

Cushion-only and min speed settings kept, as well as maxbounce.

EmeraldTiger committed:
Ogre grenade bugfix

EmeraldTiger reviewed (approved):
Declined bouncing rockets/nails, but kept cushion-only (in the form of the speed setting) and bounce counting for trigger_filter.
